File: HW1/javaApp/src/main/java/io/rienel/domain/Cat.java
package io.rienel.domain;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
public class Cat extends Animal<Mouse> {

    private final List<Mouse> mousesEaten;

    private Integer lifeCount;

    public Cat(final String name,
               final Double weight,
               final Double fullHeight,
               final Double fullLength,
               final Double fullWidth,
               final Integer lifeCount) {
        super(name, weight, AnimalType.CAT, fullHeight, fullLength, fullWidth);
        mousesEaten = new ArrayList<>();
        this.lifeCount = lifeCount;
    }

    public void dropFromTable(String item) {
        Objects.requireNonNull(item);

        System.out.println("CAT: IT'S TIME TO CRASH " + item + ".\n...\n...\nBAM... It's smacked");
    }

    public Integer getLifeCount() {
        return lifeCount;
    }

    public void setLifeCount(Integer lifeCount) {
        this.lifeCount = lifeCount;
    }

    public List<Mouse> getMousesEaten() {
        return mousesEaten;
    }

    @Override
    public void setName(String name) {
        Objects.requireNonNull(name);
        super.setName(name);
    }

    @Override
    public void setWeight(Double weight) {
        Objects.requireNonNull(weight);
        if (weight < 2.0 || weight > 10.0) {
            throw new IllegalArgumentException("Illegal weight");
        }
        super.setWeight(weight);
    }

    @Override
    public void setFullHeight(Double fullHeight) {
        Objects.requireNonNull(fullHeight);
        if (fullHeight < 2.0 || fullHeight > 10.0) {
            throw new IllegalArgumentException("Illegal fullHeight");
        }
        super.setFullHeight(fullHeight);
    }

    @Override
    public void setFullLength(Double fullLength) {
        Objects.requireNonNull(fullLength);
        if (fullLength < 2.0 || fullLength > 10.0) {
            throw new IllegalArgumentException("Illegal fullLength");
        }
        super.setFullLength(fullLength);
    }

    @Override
    public void setFullWidth(Double fullWidth) {
        Objects.requireNonNull(fullWidth);
        if (fullWidth < 2.0 || fullWidth > 10.0) {
            throw new IllegalArgumentException("Illegal fullWidth");
        }
        super.setFullWidth(fullWidth);
    }

    @Override
    public Boolean eat(Mouse food) {
        Objects.requireNonNull(food);

        mousesEaten.add(food);
        return true;
    }

    @Override
    public String speak() {
        return "CAT: MEOW";
    }
}

File: HW1/kotlinApp/src/main/kotlin/io/rienel/domain/Cat.kt
package io.rienel.domain

import io.rienel.domain.interfaces.Feedable
import io.rienel.domain.interfaces.Speekable
import java.util.StringJoiner

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
class Cat(
    name: String,
    weight: Double,
    fullHeight: Double,
    fullLength: Double,
    fullWidth: Double,
    var lifeCount: Int,
    val mousesEaten: MutableList<Mouse>
) : Animal(
    name,
    weight,
    AnimalType.CAT,
    fullHeight,
    fullLength,
    fullWidth
), Feedable<Mouse>, Speekable {

    override var name: String
        get() = super.name
        set(value) {
            assert(value.isNotEmpty())
            super.name = value
        }
    override var weight: Double
        get() = super.weight
        set(value) {
            assert(value in 2.0 .. 10.0)
            super.weight = value
        }

    override var fullHeight: Double
        get() = super.fullHeight
        set(value) {
            assert(fullHeight in 50.0 .. 100.0)
            super.fullHeight = value
        }

    override var fullLength: Double
        get() = super.fullLength
        set(value) {
            assert(fullLength in 50.0 .. 100.0)
            super.fullLength = value
        }
    override var fullWidth: Double
        get() = super.fullWidth
        set(value) {
            assert(fullWidth in 50.0 .. 100.0)
            super.fullWidth = value
        }

    fun dropFromTable(item: String?) {
        println("CAT: IT'S TIME TO CRASH $item.\n...\n...\nBAM... It's smacked")
    }

    override fun eat(food: Mouse): Boolean {
        mousesEaten.add(food)
        return true
    }

    override fun speak(): String {
        val stringJoiner = StringJoiner(" ")
        for (i in 0 until lifeCount) {
            stringJoiner.add("MEOW")
        }
        return stringJoiner.toString()
    }
}